atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraMutationTimestampProvider.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra;
+
+import com.google.common.collect.Multimap;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+
+/**
+ * This interface produces Cassandra timestamps for mutations, where it makes sense for these to be overridden.
+ * Generally, to ensure that Cassandra is able to compact tombstones smoothly, these values should increase over
```

Comment:
Probably worth noting the stronger constraints we have around the ordering of actions and the ordering of the C* timestamps that they appear at, since that constraint implies much more than using a relatively recent timestamp.

Reply:
I've added a more detailed discussion of the constraints (discussed offline).